### PR TITLE
Upload changes - changing updated time and release notes

### PR DESCRIPTION
**Pull request by ohaim1008**
***This is the initial version - we are currently testing it and adding unitests.***

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/33103

## Description
We had couple of issues with the upload flow:
* The last RN was always modified to today, due to recent changes it was uploaded daily.
* The `updated` field in the `metadata` was also modified to today.

## Minimum version of Demisto
- [ ] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Tests/Marketplace/marketplace_services.py [hunk 1]
```diff
@@ -1130,19 +1141,23 @@ def assert_upload_bucket_version_matches_release_notes_version(self,
...
+            pack_was_modified (bool): whether the pack modified or not.
+
         Returns:
             bool: whether the operation succeeded.
             bool: whether running build has not updated pack release notes.
+            str: packs latest release notes
         """
         task_status = False
         not_updated_build = False
+        packs_latest_release_notes = Pack.PACK_INITIAL_VERSION
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1160]:
```suggestion
        packs_latest_release_notes_version = Pack.PACK_INITIAL_VERSION
```
I think its better to add a version suffix as this var contains the version and not the actual rn.

**Review by guyfreund — CHANGES_REQUESTED** [review]
Did a shallow review, looking good overall.
Make sure to test the following cases:
- New Pack with RN 1.0.0
- New Pack without RN at all (empty/non existing ReleaseNotes dir)
- Update of a pack with RN bump
- Update of a pack without RN bump (readme file for example)
- Force push of a pack

File: Tests/Marketplace/marketplace_services.py [hunk 2]
```diff
@@ -1178,13 +1195,14 @@ def prepare_release_notes(self, index_folder_path, build_number):
                                                                              build_number=build_number,
                                                                              new_version=True)
 
-                        changelog[latest_release_notes] = version_changelog
+                        if version_changelog:
```
[the remark this answers is not in the document]
Comment (ohaim1008) [reply Tests/Marketplace/marketplace_services.py:1199]: The changelog isn't really empty, what happened before is that we rewrote the last changelog but if the file didn't modify there is no need to rewrite it.

File: Tests/Marketplace/marketplace_services.py [hunk 3]
```diff
@@ -1512,27 +1533,56 @@ def format_metadata(self, user_metadata, pack_content_items, integration_images,
...
+            sorted_changelog = collections.OrderedDict(sorted(changelog.items()))
+            for version, _ in sorted_changelog.items():
+                initial_changelog_version = changelog.get(version, {})
+                break
+
         init_changelog_released_date = initial_changelog_version.get('released',
                                                                      datetime.utcnow().strftime(Metadata.DATE_FORMAT))
         return init_changelog_released_date
 
+    def _get_pack_update_date(self, index_folder_path, packs_latest_release_notes):
```
Comment (yuvalbenshalom) [inline Tests/Marketplace/marketplace_services.py:1572]: @ohaim1008 @guyfreund just to make sure, If we update the readme and the pack is updated without a change in change-log, do we count this as a an update or not?

File: Tests/Marketplace/Tests/marketplace_services_test.py [hunk 4]
```diff
@@ -691,10 +751,34 @@ def test_handle_pack_create_date_changelog_exist(self, mocker, dummy_pack, is_ch
...
+        """
+           Given:
+               - existing changelog with 2 versions
+               - not existing changelog, datetime.utcnow
+           When:
+               - changelog entry already exists
+               - changelog entry not exists
+           Then:
+           - return the released field from the changelog file
+           - return datetime.utcnow
```
Comment (yuvalbenshalom) [inline Tests/Marketplace/Tests/marketplace_services_test.py:773]: I did not understand this test documentation. let's talk about it.

**Review by yuvalbenshalom — COMMENTED** [review]
see my comments

File: Tests/Marketplace/marketplace_services.py (hunk 3 above)
Comment (guyfreund) [reply Tests/Marketplace/marketplace_services.py:1572]: @yuvalbenshalom as for the current upload logic, every change in one of the pack's files is detected as an update.

**Comment by ohaim1008** [comment]
Made some changes, please look at the issue for the more information about the checked cases.

File: Tests/Marketplace/marketplace_services.py [hunk 5]
```diff
@@ -1512,27 +1533,60 @@ def format_metadata(self, user_metadata, pack_content_items, integration_images,
...
+        Args:
+            index_folder_path (str): downloaded index folder directory path.
+            pack_was_modified (bool): whether the pack was modified or not.
+        Returns:
+            datetime: Pack update date.
+        """
+        changelog = self._get_changelog(index_folder_path)
+        latest_changelog_released_date = datetime.utcnow().strftime(Metadata.DATE_FORMAT)
+
+        if changelog.keys() and not pack_was_modified:
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1583]:
```suggestion
        if changelog and not pack_was_modified:
```

File: Tests/Marketplace/marketplace_services.py [hunk 6]
```diff
@@ -1512,27 +1533,60 @@ def format_metadata(self, user_metadata, pack_content_items, integration_images,
...
+            packs_latest_release_notes = (list(reversed(sorted(changelog.keys()))))[0]
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1584]:
```suggestion
            packs_latest_release_notes = max(LooseVersion(ver) for ver in changelog)
```

File: Tests/Marketplace/upload_packs.py [hunk 7]
```diff
@@ -874,18 +874,27 @@ def main():
...
-                                           packs_statistic_df=packs_statistic_df)
+                                           packs_statistic_df=packs_statistic_df,
+                                           pack_was_modified=pack_was_modified)
         if not task_status:
             pack.status = PackStatus.FAILED_METADATA_PARSING.name
             pack.cleanup()
             continue
 
-        task_status, not_updated_build = pack.prepare_release_notes(index_folder_path, build_number)
+        task_status, not_updated_build, packs_latest_release_notes = \
```
Comment (guyfreund) [inline Tests/Marketplace/upload_packs.py:896]:
```suggestion
        task_status, not_updated_build = \
```
`packs_latest_release_notes` isn't of use, let's remove it from here and from the `prepare_rn` function.

File: Tests/Marketplace/marketplace_services.py (hunk 5 above)
Comment (guyfreund) [reply Tests/Marketplace/marketplace_services.py:1583]: In what cases `changelog` can be empty? not sure why what's the right behavior in that case

File: Tests/Marketplace/marketplace_services.py [hunk 8]
```diff
@@ -1512,27 +1533,60 @@ def format_metadata(self, user_metadata, pack_content_items, integration_images,
...
 
+        changelog = self._get_changelog(index_folder_path)
         initial_changelog_version = changelog.get(Pack.PACK_INITIAL_VERSION, {})
+
+        if not initial_changelog_version:
+            # if the changelog exists but the first version isn't PACK_INITIAL_VERSION
+            sorted_changelog = collections.OrderedDict(sorted(changelog.items()))
+            for version, _ in sorted_changelog.items():
+                initial_changelog_version = changelog.get(version, {})
+                break
```
Comment (guyfreund) [inline Tests/Marketplace/marketplace_services.py:1566]:
```suggestion
            initial_changelog_version = min(LooseVersion(ver) for ver in changelog)
```

**Review by guyfreund — CHANGES_REQUESTED** [review]

File: Tests/Marketplace/marketplace_services.py (hunk 5 above)
Comment (amshamah419) [reply Tests/Marketplace/marketplace_services.py:1583]: If it's a new pack, then it _could_ be empty.

Comment (guyfreund) [reply Tests/Marketplace/marketplace_services.py:1583]: But then `pack_was_modified` will be set to True, no?

Comment (amshamah419) [reply Tests/Marketplace/marketplace_services.py:1583]: I don't believe so. `pack_was_modified` should only be used when a pack is being re-uploaded. A new pack is a fresh commit so there wouldn't be an existing version to modify

**Review by guyfreund — APPROVED** [review]
Looks good, please trigger an upload before merging this

**Review by yuvalbenshalom — APPROVED** [review]

**Pull request merged**